sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1358 @@
+ *      .withParametersMapper( (row, map ) -> {
+ *        map.put("id", row.getString("id));
+ *        map.put("first", row.getString("firstName"));
+ *        map.put("last", row.getString("lastName"));
+ *      })
+ *    );
+ * }</pre>
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class Neo4jIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Neo4jIO.class);
+
+  /**
+   * Read all rows using a Neo4j Cypher query.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   * @param <OutputT> Type of the data to be read.
+   */
+  public static <ParameterT, OutputT> Read<ParameterT, OutputT> read() {
+    return new AutoValue_Neo4jIO_Read.Builder<ParameterT, OutputT>()
+        .setFetchSize(ValueProvider.StaticValueProvider.of(Config.defaultConfig().fetchSize()))
+        .build();
+  }
+
+  /**
+   * Write all rows using a Neo4j Cypher UNWIND cypher statement. This sets a default batch batch
+   * size of 5000.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   */
+  public static <ParameterT> WriteUnwind<ParameterT> writeUnwind() {
+    return new AutoValue_Neo4jIO_WriteUnwind.Builder<ParameterT>()
+        .setBatchSize(ValueProvider.StaticValueProvider.of(5000L))
+        .build();
+  }
+
+  private static <ParameterT, OutputT> PCollection<OutputT> getOutputPCollection(
+      PCollection<ParameterT> input, DoFn<ParameterT, OutputT> writeFn, Coder<OutputT> coder) {
+    PCollection<OutputT> output = input.apply(ParDo.of(writeFn)).setCoder(coder);
+
+    try {
+      TypeDescriptor<OutputT> typeDesc = coder.getEncodedTypeDescriptor();
+      SchemaRegistry registry = input.getPipeline().getSchemaRegistry();
+      Schema schema = registry.getSchema(typeDesc);
+      output.setSchema(
+          schema,
+          typeDesc,
+          registry.getToRowFunction(typeDesc),
+          registry.getFromRowFunction(typeDesc));
+    } catch (NoSuchSchemaException e) {
+      // ignore
+    }
+    return output;
+  }
+
+  /**
+   * An interface used by {@link Read} for converting each row of a Neo4j {@link Result} record
+   * {@link Record} into an element of the resulting {@link PCollection}.
+   */
+  @FunctionalInterface
+  public interface RowMapper<T> extends Serializable {
+    T mapRow(Record record) throws Exception;
+  }
+
+  /**
+   * An interface used by {@link Read} for converting input parameter data to a parameters map which
+   * can be used by your Neo4j cypher query.
+   */
+  @FunctionalInterface
+  public interface ParametersMapper<ParameterT> extends Serializable {
+    /**
+     * Simply map an input record to a parameters map.
+     *
+     * @param input the input read
+     * @param parametersMap the parameters map to update in this llabmda
+     */
+    void mapParameters(ParameterT input, Map<String, Object> parametersMap);
```

Comment:
Looking at the code I was wondering if this mapper needs to be stateful? 
Can't this just be a function `ParameterT ->  Map<String, Object>`?

Reply:
This idea was also taken from JdbcIO where a PreparedStatementSetter is functionally nearly identical except that values are not set in a PreparedStatement object related to JDBC code but actually in a Map<String, Object> which is passed to the Neo4j along with the Cypher statement in question. 
That being said, I think you are in fact correct.  For the readAll() scenario every input element causes exactly one Cypher statement to be executed and exactly one set of parameters to be mapped so it is a function.
It's not that hard to change to so I'll just do that.

Reply:
The recommended change from ParametersMapper to SerializableFunction was done in commit [a257cec](https://github.com/apache/beam/pull/15916/commits/a257ceced20d132f7a10e3ebcbd260f7b0fb4a2d)